### ISSUE TITLE
Error when running error prone 2.0.12 on AbstractList from AOSP

### ISSUE THREAD

**Issue by paulduffin** [google/error-prone#468]
The source code can be obtained from https://source.android.com/source/downloading.html. Running error prone is more tricky but I if you email me I can talk you through it.

I am using version 1.9.0-dev-r2973-2 of javac that comes with it.

libcore/ojluni/src/main/java/java/util/AbstractList.java:71: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
public abstract class AbstractList<E> extends AbstractCollection<E> implements List<E> {
                ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:

```
 error-prone version: 2.0.12
 Stack Trace:
 java.lang.NullPointerException
    at com.sun.tools.javac.comp.Resolve$AbstractMethodCheck.argumentsAcceptable(Resolve.java:717)
    at com.sun.tools.javac.comp.Resolve$4.argumentsAcceptable(Resolve.java:835)
    at com.sun.tools.javac.comp.Resolve.rawInstantiate(Resolve.java:576)
    at com.sun.tools.javac.comp.Resolve.selectBest(Resolve.java:1440)
    at com.sun.tools.javac.comp.Resolve.findMethodInScope(Resolve.java:1621)
    at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1690)
    at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1664)
    at com.sun.tools.javac.comp.Resolve$9.doLookup(Resolve.java:2364)
    at com.sun.tools.javac.comp.Resolve$BasicLookupHelper.lookup(Resolve.java:2972)
    at com.sun.tools.javac.comp.Resolve.lookupMethod(Resolve.java:3223)
    at com.sun.tools.javac.comp.Resolve.resolveQualifiedMethod(Resolve.java:2361)
    at com.sun.tools.javac.comp.Resolve.resolveInternalMethod(Resolve.java:2429)
    at com.google.errorprone.bugpatterns.EqualsHashCode.resolveMethod(EqualsHashCode.java:121)
    at com.google.errorprone.bugpatterns.EqualsHashCode.matchClass(EqualsHashCode.java:97)
    at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:629)
    at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:144)
    at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:778)
    at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:72)
    at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
    at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
    at com.sun.source.util.TreeScanner.scan(TreeScanner.java:106)
    at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:114)
    at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:145)
    at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:650)
    at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:144)
    at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:561)
    at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:53)
    at com.google.errorprone.scanner.Scanner.scan(Scanner.java:68)
    at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
    at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:123)
    at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:113)
    at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1303)
    at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1247)
    at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:858)
    at com.sun.tools.javac.main.Main.compile(Main.java:253)
    at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:187)
    at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:98)
    at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:128)
    at com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:70)
    at com.google.errorprone.ErrorProneCompiler.main(ErrorProneCompiler.java:49)
```

**Comment by paulduffin** [comment]
Tried with v2.0.13 and got identical problem even down to line numbers.

libcore/ojluni/src/main/java/java/util/AbstractList.java:71: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
public abstract class AbstractList<E> extends AbstractCollection<E> implements List<E> {
                ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.0.13
     Stack Trace:
     java.lang.NullPointerException
  	at com.sun.tools.javac.comp.Resolve$AbstractMethodCheck.argumentsAcceptable(Resolve.java:717)
  	at com.sun.tools.javac.comp.Resolve$4.argumentsAcceptable(Resolve.java:835)
  	at com.sun.tools.javac.comp.Resolve.rawInstantiate(Resolve.java:576)
  	at com.sun.tools.javac.comp.Resolve.selectBest(Resolve.java:1440)
  	at com.sun.tools.javac.comp.Resolve.findMethodInScope(Resolve.java:1621)
  	at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1690)
  	at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1664)
  	at com.sun.tools.javac.comp.Resolve$9.doLookup(Resolve.java:2364)
  	at com.sun.tools.javac.comp.Resolve$BasicLookupHelper.lookup(Resolve.java:2972)
  	at com.sun.tools.javac.comp.Resolve.lookupMethod(Resolve.java:3223)
  	at com.sun.tools.javac.comp.Resolve.resolveQualifiedMethod(Resolve.java:2361)
  	at com.sun.tools.javac.comp.Resolve.resolveInternalMethod(Resolve.java:2429)
  	at com.google.errorprone.bugpatterns.EqualsHashCode.resolveMethod(EqualsHashCode.java:121)
  	at com.google.errorprone.bugpatterns.EqualsHashCode.matchClass(EqualsHashCode.java:97)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:629)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:144)
  	at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:778)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:72)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
  	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:106)
  	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:114)
  	at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:145)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:650)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:144)
  	at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:561)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:53)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:68)
  	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
  	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:123)
  	at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:113)
  	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1303)
  	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1247)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:858)
  	at com.sun.tools.javac.main.Main.compile(Main.java:253)
  	at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:187)
  	at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:98)
  	at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:128)
  	at com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:70)
  	at com.google.errorprone.ErrorProneCompiler.main(ErrorProneCompiler.java:49)
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

**Comment by cushon** [comment]
If you get a chance, can you let us know if this is still happening with 2.0.16? That release is using a newer version of javac that includes a bunch of fixes.

**Comment by paulduffin** [comment]
I will do but I won't be able to test it for a while.

**Comment by paulduffin** [comment]
Just had a quick look and cannot find a prebuilt version of 2.0.16. Do you
know where I can find one (it is not yet up on
https://mvnrepository.com/artifact/com.google.errorprone)? It will be a lot
quicker to test if I do not have to build from source.

**Comment by cushon** [comment]
You can get the jars from maven central:
http://repo1.maven.org/maven2/com/google/errorprone/error_prone_ant/2.0.16/

**Comment by paulduffin** [comment]
Thanks.

**Comment by paulduffin** [comment]
What version of javac do I need? 9-dev-r3297-3/ ?

**Comment by cushon** [comment]
Yep, `9-dev-r3297-3`.

**Comment by paulduffin** [comment]
I get the following crashes when I try and build Android with it. It does
parallel compilation which is why I get more than one failure and I
actually get two versions of the second crash.

     error-prone version: 2.0.16
     Stack Trace:
     java.lang.NullPointerException
  at com.sun.tools.javac.tree.TreeInfo.skipParens(TreeInfo.java:708)
  at com.google.errorprone.util.ASTHelpers.constValue(ASTHelpers.java:680)
  at com.google.errorprone.util.ASTHelpers.constValue(ASTHelpers.java:699)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitForLoop(Reachability.java:298)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitForLoop(Reachability.java:62)
  at com.sun.tools.javac.tree.JCTree$JCForLoop.accept(JCTree.java:1102)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.scan(Reachability.java:82)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.scan(Reachability.java:76)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitBlock(Reachability.java:101)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitBlock(Reachability.java:62)
  at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:996)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.scan(Reachability.java:82)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitIf(Reachability.java:150)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitIf(Reachability.java:62)
  at com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1397)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.scan(Reachability.java:82)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.scan(Reachability.java:76)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitBlock(Reachability.java:101)
  at
com.google.errorprone.util.Reachability$CanCompleteNormallyVisitor.visitBlock(Reachability.java:62)
  at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:996)
  at
com.google.errorprone.util.Reachability.canCompleteNormally(Reachability.java:59)
  at
com.google.errorprone.bugpatterns.FallThrough.matchSwitch(FallThrough.java:69)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitSwitch(ErrorProneScanner.java:1075)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitSwitch(ErrorProneScanner.java:144)
  at com.sun.tools.javac.tree.JCTree$JCSwitch.accept(JCTree.java:1193)
  at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
  at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:575)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:144)
  at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:996)
  at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
  at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:943)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:144)
  at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:868)
  at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
  at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:635)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:144)
  at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:778)
  at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
  at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  at
com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:650)
  at
com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:144)
  at
com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:561)
  at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:52)
  at com.google.errorprone.scanner.Scanner.scan(Scanner.java:68)
  at
com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
  at
com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:145)
  at
com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
  at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1277)
  at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1226)
  at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:840)
  at com.sun.tools.javac.main.Main.compile(Main.java:257)
  at
com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:229)
  at
com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:107)
  at
com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:128)
  at
com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:70)
  at
com.google.errorprone.ErrorProneCompiler.main(ErrorProneCompiler.java:49)

     java.lang.NullPointerException
        at
com.sun.tools.javac.comp.Resolve$AbstractMethodCheck.treeForDiagnostics(Resolve.java:764)
        at
com.sun.tools.javac.comp.Resolve$AbstractMethodCheck.argumentsAcceptable(Resolve.java:724)
        at
com.sun.tools.javac.comp.Resolve$4.argumentsAcceptable(Resolve.java:848)
        at com.sun.tools.javac.comp.Resolve.rawInstantiate(Resolve.java:582)
        at com.sun.tools.javac.comp.Resolve.selectBest(Resolve.java:1503)
        at
com.sun.tools.javac.comp.Resolve.findMethodInScope(Resolve.java:1684)
        at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1753)
        at com.sun.tools.javac.comp.Resolve.findMethod(Resolve.java:1727)
        at com.sun.tools.javac.comp.Resolve$9.doLookup(Resolve.java:2433)
        at
com.sun.tools.javac.comp.Resolve$BasicLookupHelper.lookup(Resolve.java:3041)
        at com.sun.tools.javac.comp.Resolve.lookupMethod(Resolve.java:3291)
        at
com.sun.tools.javac.comp.Resolve.resolveQualifiedMethod(Resolve.java:2430)
        at
com.sun.tools.javac.comp.Resolve.resolveInternalMethod(Resolve.java:2500)
        at
com.google.errorprone.util.ASTHelpers.resolveExistingMethod(ASTHelpers.java:902)
        at
com.google.errorprone.bugpatterns.EqualsHashCode.matchClass(EqualsHashCode.java:90)
        at
com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:629)
        at
com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:144)
        at
com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:778)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at
com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at
com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
        at
com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:650)
        at
com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:144)
        at
com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:561)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:52)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:68)
        at
com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at
com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:145)
        at
com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
        at
com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1277)
        at
com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1226)
        at
com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:840)
        at com.sun.tools.javac.main.Main.compile(Main.java:257)
        at
com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:229)
        at
com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:107)
        at
com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:128)
        at
com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:70)
        at
com.google.errorprone.ErrorProneCompiler.main(ErrorProneCompiler.java:49)

**Comment by cushon** [comment]
```
     java.lang.NullPointerException
  at com.sun.tools.javac.tree.TreeInfo.skipParens(TreeInfo.java:708)
  at com.google.errorprone.util.ASTHelpers.constValue(ASTHelpers.java:680)
  at com.google.errorprone.util.ASTHelpers.constValue(ASTHelpers.java:699)
```

This was fixed by 3018c0dfa2f5a1e9e7d1c1739d354073063d0f85.

I'm not sure about the second one yet.

**Comment by cushon** [comment]
```
     java.lang.NullPointerException
        at
com.sun.tools.javac.comp.Resolve$AbstractMethodCheck.treeForDiagnostics(Resolve.java:764)
        at
...
        at
com.google.errorprone.util.ASTHelpers.resolveExistingMethod(ASTHelpers.java:902)
        at
com.google.errorprone.bugpatterns.EqualsHashCode.matchClass(EqualsHashCode.java:90)
```

The workaround for this one is `-XDcompilePolicy=simple`. ce3f1f0e4959614c9a61a9e1b5a7b5cd70678524 was supposed to address this but it isn't a complete fix.

We should think about baking the 'simple' policy into all Error Prone javac invocations, or applying the fix from the thread below to `error-prone-javac`.

previously: http://mail.openjdk.java.net/pipermail/compiler-dev/2016-December/010568.html
